src/typecheck.cpp

```diff
@@ -240,10 +240,12 @@ void typecheck(TypedAST::RecordAccessExpression* ast, TypeChecker& tc) {
 
 void typecheck(TypedAST::ConstructorExpression* ast, TypeChecker& tc) {
 	typecheck(ast->m_constructor, tc);
-	TypeFunctionId tf = tc.m_core.m_tf_core.find_term(ast->m_constructor->m_value_type);
-	TypeFunctionData& tf_data = tc.m_core.m_type_functions[tf];
 
-	// TODO: match type arguments as well, we only support 0 now
+	auto handle = static_cast<TypedAST::MonoTypeHandle*>(ast->m_constructor);
+	assert(handle && "Error: Value constructor was not handled");
```

Comment:
static cast doesn't return null if you perform an runtime-invalid cast. it wrongly succeedes without warning.

did you mean to `assert(ast->m_constructor->type() == TypedASTTag::MonoTypeHandle)` ?